### PR TITLE
Change MATLAB format to accept comma delimiter

### LINKED ISSUE #2277: Support commas between entries in pl-matrix-input MATLAB format
pl-matrix-input MATLAB format currently wants something like:
```
[3 4; 5 6]
```

We should also accept this with commas, like:
```
[3, 4; 5, 6]
```

### PR DESCRIPTION
**Pull request by tdy**
Fixes #2277 

`string_to_2darray` now matches the MATLAB console:
```matlab
[1, 2, 3; 4, 5, 6]  %=[1 2 3; 4 5 6]
```
```matlab
[, 1    , 2  , 3; 4   ,5, 6 ,]  %=[1 2 3; 4 5 6]
```
~~except with repeated commas (MATLAB rejects, `string_to_2darray` accepts):~~
```matlab
[1,,,,, 2, 3; 4, 5, 6]  %=now correctly rejected
```

---

* [x] Add basic support
* [x] Fix edge cases for MATLAB compat
* [x] Update help text (`?` and `Why?`)
* [x] Fix "invalid" case in element tests
* [x] Add more "invalid" tests
* [x] Add css to help text and error string ~~(but not the error string -- that won't be trivial)~~
* [x] Change to f-strings
* [x] Reword ambiguous `Entry (1, 1) of matrix "a"` to `Entry "a" at (1, 1) of matrix`
* [x] Escape invisible chars

### CODE REVIEW

**Comment by coatless** [comment]
I'm seeing:

```
            "stack": "Error: \"format_errors keys\" mismatch: expected \"[\"C\"]\" but got \"[]\"\n    at checkEqual (/PrairieLearn/lib/question.js:181:488)\n    at Object._compareSubmissions (/PrairieLearn/lib/question.js:181:1289)\n    at /PrairieLearn/lib/question.js:192:3145\n    at /PrairieLearn/node_modules/async/dist/async.js:2948:28\n    at replenish (/PrairieLearn/node_modules/async/dist/async.js:440:21)\n    at iterateeCallback (/PrairieLearn/node_modules/async/dist/async.js:424:21)\n    at /PrairieLearn/node_modules/async/dist/async.js:321:20\n    at /PrairieLearn/node_modules/async/dist/async.js:2953:17\n    at /PrairieLearn/lib/question.js:192:3057\n    at /PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:622:9\n    at Query.<anonymous> (/PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:225:13)\n    at Query.handleReadyForQuery (/PrairieLearn/node_modules/pg/lib/query.js:121:12)\n    at Connection.<anonymous> (/PrairieLearn/node_modules/pg/lib/client.js:256:19)\n    at Connection.emit (events.js:310:20)\n    at Connection.EventEmitter.emit (domain.js:482:12)\n    at Socket.<anonymous> (/PrairieLearn/node_modules/pg/lib/connection.js:120:12)\n    at Socket.emit (events.js:310:20)\n    at Socket.EventEmitter.emit (domain.js:482:12)\n    at addChunk (_stream_readable.js:286:12)\n    at readableAddChunk (_stream_readable.js:268:9)\n    at Socket.Readable.push (_stream_readable.js:209:10)\n    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)"
```

https://github.com/PrairieLearn/PrairieLearn/runs/600347802#step:9:14093

**Comment by tdy** [comment]
> 
> 
> I'm seeing:
> 
> ```
>             "stack": "Error: \"format_errors keys\" mismatch: expected \"[\"C\"]\" but got \"[]\"\n    at checkEqual (/PrairieLearn/lib/question.js:181:488)\n    at Object._compareSubmissions (/PrairieLearn/lib/question.js:181:1289)\n    at /PrairieLearn/lib/question.js:192:3145\n    at /PrairieLearn/node_modules/async/dist/async.js:2948:28\n    at replenish (/PrairieLearn/node_modules/async/dist/async.js:440:21)\n    at iterateeCallback (/PrairieLearn/node_modules/async/dist/async.js:424:21)\n    at /PrairieLearn/node_modules/async/dist/async.js:321:20\n    at /PrairieLearn/node_modules/async/dist/async.js:2953:17\n    at /PrairieLearn/lib/question.js:192:3057\n    at /PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:622:9\n    at Query.<anonymous> (/PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:225:13)\n    at Query.handleReadyForQuery (/PrairieLearn/node_modules/pg/lib/query.js:121:12)\n    at Connection.<anonymous> (/PrairieLearn/node_modules/pg/lib/client.js:256:19)\n    at Connection.emit (events.js:310:20)\n    at Connection.EventEmitter.emit (domain.js:482:12)\n    at Socket.<anonymous> (/PrairieLearn/node_modules/pg/lib/connection.js:120:12)\n    at Socket.emit (events.js:310:20)\n    at Socket.EventEmitter.emit (domain.js:482:12)\n    at addChunk (_stream_readable.js:286:12)\n    at readableAddChunk (_stream_readable.js:268:9)\n    at Socket.Readable.push (_stream_readable.js:209:10)\n    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)"
> ```
> 
> https://github.com/PrairieLearn/PrairieLearn/runs/600347802#step:9:14093

Yeah, not sure what that was about.

I triggered a retest (close/reopen PR) and it passed on the retest. My local `PrairieLearn/docker/tests` have passed. I think it was just a false alarm?

**Comment by tdy** [comment]
And my innocuous `exam02` PR just failed CI on a retest. I think Travis CI is just acting up today.

File: question-servers/freeformPythonLib/prairielearn.py [hunk 1]
```diff
@@ -776,8 +772,8 @@ def string_to_2darray(s, allow_complex=True):
         if (m == 0):
             return (None, {'format_error': 'Matrix has no rows.'})
 
-        # Get number of columns by splitting first row on space
-        n = len(s[0].split())
+        # Get number of columns by splitting first row on space (treat comma as space)
+        n = len(s[0].replace(',', ' ').split())
```
Comment (tdy) [inline question-servers/freeformPythonLib/prairielearn.py:776]: I'm thinking now that the `replace`s can be refactored to lines 766 and 938 (but haven't tested).

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py]: Done+tested on my end

File: question-servers/freeformPythonLib/prairielearn.py [hunk 2]
```diff
@@ -762,12 +762,8 @@ def string_to_2darray(s, allow_complex=True):
...
         #       is '[1 - 2j]' the same as '[1 -2j]' or '[1-2j]'
 
-        # Return error if there are any commas
-        if ',' in s:
-            return (None, {'format_error': 'Commas cannot be used as delimiters in an expression with single brackets.'})
-
-        # Split on semicolon
-        s = s.split(';')
+        # Split on semicolon (treat comma as space)
+        s = s.replace(',', ' ').split(';')
```
Comment (mwest1066) [inline question-servers/freeformPythonLib/prairielearn.py:766]: Would it be more robust to do check whether the entire original string contains a comma and, if so, split all rows on comma and otherwise split on space? That would reject `[1,,,,, 2, 3; 4, 5, 6]`.

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py:766]: Without adding more complicated logic, that would also reject stuff like `[1, 2, 3; 4 5 6]` (mixing spaces and commas which MATLAB allows). Which limitation do you prefer?

Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py:766]: Oh! I didn't realize Matlab would accept that. How horrible! 😄 

In that case, can't you first split on semi-colon, then for each row check whether it contains a comma and split on comma or space? I agree this is slightly more complex than your current code, but you could write it like:
```
n = len(s[0].split(',' if ',' in s[0] else None))

s_row = s[i].split(,' if ',' in s[0] else None)
```

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py:766]: So my original counter-example didn't make this clear (edited later), but MATLAB allows mixing commas/spaces even within rows. I can try to make PL's parser mimic MATLAB, but I wasn't sure if that's where we wanted to go with it. Sounds like that's what you prefer?

**Comment by nicknytko** [comment]
> > I'm seeing:
> > ```
> >             "stack": "Error: \"format_errors keys\" mismatch: expected \"[\"C\"]\" but got \"[]\"\n    at checkEqual (/PrairieLearn/lib/question.js:181:488)\n    at Object._compareSubmissions (/PrairieLearn/lib/question.js:181:1289)\n    at /PrairieLearn/lib/question.js:192:3145\n    at /PrairieLearn/node_modules/async/dist/async.js:2948:28\n    at replenish (/PrairieLearn/node_modules/async/dist/async.js:440:21)\n    at iterateeCallback (/PrairieLearn/node_modules/async/dist/async.js:424:21)\n    at /PrairieLearn/node_modules/async/dist/async.js:321:20\n    at /PrairieLearn/node_modules/async/dist/async.js:2953:17\n    at /PrairieLearn/lib/question.js:192:3057\n    at /PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:622:9\n    at Query.<anonymous> (/PrairieLearn/node_modules/@prairielearn/prairielib/lib/sql-db.js:225:13)\n    at Query.handleReadyForQuery (/PrairieLearn/node_modules/pg/lib/query.js:121:12)\n    at Connection.<anonymous> (/PrairieLearn/node_modules/pg/lib/client.js:256:19)\n    at Connection.emit (events.js:310:20)\n    at Connection.EventEmitter.emit (domain.js:482:12)\n    at Socket.<anonymous> (/PrairieLearn/node_modules/pg/lib/connection.js:120:12)\n    at Socket.emit (events.js:310:20)\n    at Socket.EventEmitter.emit (domain.js:482:12)\n    at addChunk (_stream_readable.js:286:12)\n    at readableAddChunk (_stream_readable.js:268:9)\n    at Socket.Readable.push (_stream_readable.js:209:10)\n    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)"
> > ```
> > 
> > 
> > https://github.com/PrairieLearn/PrairieLearn/runs/600347802#step:9:14093
> 
> Yeah, not sure what that was about.
> 
> I triggered a retest (close/reopen PR) and it passed on the retest. My local `PrairieLearn/docker/tests` have passed. I think it was just a false alarm?

I would check out the `test()` function in `pl-matrix-input.py`, it might be that being more forgiving in the inputs the element takes will break tests for invalid submissions.

File: question-servers/freeformPythonLib/prairielearn.py (hunk 2 above)
Comment (nicknytko) [reply question-servers/freeformPythonLib/prairielearn.py:766]: Does python allow splits with regex?  Could we just split each row whenever it literally sees a comma or a space(s)?

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py:766]: Yep if we go the true-compat route, I was planning to use `re`.

Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py:766]: I feel like I've bikeshedded this PR to death, but I guess I'd agree that we should just `re` it. Does `re.split(r'\s*(\s|,)\s*)` do what we want?

Comment (echuber2) [reply question-servers/freeformPythonLib/prairielearn.py]: I think this will break if the student has newlines in a Matlab matrix definition as `\s` matches those. So it's important to split by `;` or `\n` first. I'd need to read the Matlab specification to see if there are other edgecases. @tdy Does this cover it?

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py]: @echuber which "this" are you referencing? The old commit, matt's regex, or the newest commit?

Comment (echuber2) [reply question-servers/freeformPythonLib/prairielearn.py]: Sorry, Matt's regex in the context of this thread.

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py]: The latest revision splits into cols after first splitting into rows.

Comment (echuber2) [reply question-servers/freeformPythonLib/prairielearn.py]: I can't tell from the code whether it correctly handles the case where students had literally put a newline in their matrix entry. Does the input element prevent that? Octave allows this:

```
>> m = [1 2, 3; 4, 5 6
7 8 9]
m =

   1   2   3
   4   5   6
   7   8   9
```

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py]: I don't know about the bigger picture, but it looks like `string_to_2darray` was written solely to parse `pl-matrix-input`, which uses `<input>` and makes newlines moot.

Comment (echuber2) [reply question-servers/freeformPythonLib/prairielearn.py]: If there's a potential this could be used some context in the future where that would be a problem, maybe we should also do splitting by `\n` here then, and check that the number of columns per row is consistent (if not already doing so)

File: question-servers/freeformPythonLib/prairielearn.py [hunk 3]
```diff
@@ -789,11 +795,17 @@ def string_to_2darray(s, allow_complex=True):
...
-            # Split on space
-            s_row = s[i].split()
+            # Split row
+            s_row = re.split(matlab_delimiter_regex, s[i])
+
+            # Ignore first/last token if empty string (occurs when row leads/trails with valid delimiter)
+            if s_row and not s_row[0]:
+                s_row.pop(0)
+            if s_row and not s_row[-1]:
+                s_row.pop(-1)
```
Comment (tdy) [inline question-servers/freeformPythonLib/prairielearn.py:805]: For `re.split()` side effect:

Valid delimiter at beginning/end of input will add exactly 1 empty string to beginning/end of output (2+ if invalid)

File: elements/pl-matrix-component-input/pl-matrix-component-input.py [hunk 4]
```diff
@@ -379,7 +379,7 @@ def test(element_html, data):
                 feedback.update({each_entry_name: 'incorrect'})
             elif result == 'invalid':
                 if random.choice([True, False]):
-                    data['raw_submitted_answers'][each_entry_name] = '1,2'
+                    data['raw_submitted_answers'][each_entry_name] = '1,,2'
```
Comment (tdy) [inline elements/pl-matrix-component-input/pl-matrix-component-input.py:382]:
```suggestion
                    data['raw_submitted_answers'][each_entry_name] = '1,2'
```

File: question-servers/freeformPythonLib/prairielearn.py (hunk 2 above)
Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py]: sounds good

File: question-servers/freeformPythonLib/prairielearn.py [hunk 5]
```diff
@@ -776,8 +772,18 @@ def string_to_2darray(s, allow_complex=True):
         if (m == 0):
             return (None, {'format_error': 'Matrix has no rows.'})
 
-        # Get number of columns by splitting first row on space
-        n = len(s[0].split())
+        # Regex to split rows a la MATLAB
+        matlab_delimiter_regex = re.compile(r'\s*[\s|,]\s*')
```
Comment (mwest1066) [inline question-servers/freeformPythonLib/prairielearn.py:776]: If you are going to use the character set operator `[]` then I think you don't want the alternative operator `|`, because inside `[]` the vertical bar is interpreted as a literal character. For example, the regex as written will also match `4|5|23|2`.
```suggestion
        matlab_delimiter_regex = re.compile(r'\s*[\s,]\s*')
```

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py:776]: Oops yeah. I switched brackets because `r'\s*(\s|,)\s*'` was capturing all the delimiters, e.g. `1, 2, 3' -> ['1', ',', '2', ',', '3']`

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py]: `r'\s*[\s,]\s*'` is working on my end

Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py]: excellent!

**Comment by mwest1066** [comment]
Two reporting error issues below. I'm not sure whether these were caused by this PR or were pre-existing.

image

**Comment by mwest1066** [comment]
If there are non-printable characters in the input then they show up in the submission panel, but not in the popover help (see below). Would it be possible to call the same function that's used for the submission display to render the invalid entry? Or if that's a mess, maybe it's safest to just not show it and only say "entry (1,3) is invalid".

image

**Comment by tdy** [comment]
* [x] Reworded `Entry (1, 1) of matrix "a"` to `Entry "a" at (1, 1) of matrix`
* [x] Escaped invisible chars
* [x] Added css to help text (but not the error string -- that won't be trivial)

**Comment by mwest1066** [comment]
I love the change to f-strings! They are they best!

File: question-servers/freeformPythonLib/prairielearn.py [hunk 6]
```diff
@@ -828,7 +828,7 @@ def string_to_2darray(s, allow_complex=True):
                     A[i, j] = ans
                 except Exception:
                     # Return error if entry could not be converted to float or complex
-                    return (None, {'format_error': f'Entry ({i + 1}, {j + 1}) of matrix "{s_row[j]}" has invalid format.'})
+                    return (None, {'format_error': f'Entry "{s_row[j]}" at ({i + 1}, {j + 1}) of matrix has invalid format.'})
```
Comment (mwest1066) [inline question-servers/freeformPythonLib/prairielearn.py]: Maybe it's worth being even more explicit here? Something like the following?
```suggestion
                    return (None, {'format_error': f'Entry "{s_row[j]}" at location (row={i + 1}, column={j + 1}) in the matrix has invalid format.'})
```

File: question-servers/freeformPythonLib/prairielearn.py [hunk 7]
```diff
@@ -913,85 +925,13 @@ def string_to_2darray(s, allow_complex=True):
                     A[i, j] = ans
                 except Exception:
                     # Return error if entry could not be converted to float or complex
-                    return (None, {'format_error': 'Entry ({:d}, {:d}) of matrix "{:s}" has invalid format.'.format(i + 1, j + 1, s_row[i][j])})
+                    return (None, {'format_error': f'Entry "{escape_unicode_string(s_row[i][j])}" at ({i + 1}, {j + 1}) of matrix has invalid format.'})
```
Comment (mwest1066) [inline question-servers/freeformPythonLib/prairielearn.py:928]: I guess we should treat `format_errors` as HTML? So we could do something like this:

```suggestion
                    return (None, {'format_error': f'Entry <code class="user-output-invalid">{html.escape(escape_unicode_string(s_row[i][j]))}</code> at ({i + 1}, {j + 1}) of matrix has invalid format.'})
```

Then in the popover template in `pl-matrix-input.mustache` we would use triple-braces on `{{{format_error_message}}}`?

I guess this will require checking whether we need to put `html.escape()` around any other format error messages?

Comment (tdy) [reply question-servers/freeformPythonLib/prairielearn.py:928]: Ah so actually it is trivial after all -- I wasn't aware of `{{{`. But I also gave it up quickly because it felt a bit code-smelly to put the markup in server code 😵

Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py]: Yep, I agree with you on the code smell. But now that I think about it further I think we probably do want to be able able to put HTML in `format_errors`, because they are fundamentally for display to the student and being able to style the errors can be helpful in making them clearer. Unfortunately it means we need to be careful with escaping! 😝

File: question-servers/freeformPythonLib/prairielearn.py [hunk 8]
```diff
@@ -1135,3 +1075,12 @@ def escape_unprintable(x):
             return x
 
     return ''.join(map(escape_unprintable, string))
+
+
+def escape_invalid_string(string):
```
Comment (tdy) [inline question-servers/freeformPythonLib/prairielearn.py:1080]: Added this helper

File: question-servers/freeformPythonLib/prairielearn.py [hunk 9]
```diff
@@ -1135,3 +1075,12 @@ def escape_unprintable(x):
...
+    """
+    escape_invalid_string(string)
+
+    Wraps and escapes string in <code> tags.
+    """
+    return f'<code class="user-output-invalid">{html.escape(escape_unicode_string(string))}</code>'
```
Comment (nicknytko) [inline question-servers/freeformPythonLib/prairielearn.py:1086]: Nice, I like keeping the gross `<code>` tags out of the actual code.

File: elements/pl-matrix-input/pl-matrix-input.py [hunk 10]
```diff
@@ -313,10 +313,50 @@ def test(element_html, data):
...
+                ],
+                'empty matrix': [
+                    '[]',
+                    '[,]',
+                ],
+            }
+
+            error = random.choice(list(invalid_cases))
+            data['raw_submitted_answers'][name] = random.choice(invalid_cases[error])
+            data['format_errors'][name] = error
```
Comment (nicknytko) [inline elements/pl-matrix-input/pl-matrix-input.py:359]: Thanks, these tests look great.

File: question-servers/freeformPythonLib/prairielearn.py (hunk 8 above)
Comment (mwest1066) [reply question-servers/freeformPythonLib/prairielearn.py:1080]: Excellent thinking!
